File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	bytesToSend := len(bytes)
	bytesSent := 0

	for bytesSent < bytesToSend {
		n, err := socket.Write(bytes[bytesSent:])
		if err != nil {
			return err
		}
		bytesSent += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	bytesReceived := 0
	for bytesReceived < size {
		n, err := socket.Read(buff[bytesReceived:])
		if err != nil {
			return nil, err
		}
		bytesReceived += n
	}
	return buff, nil
}

File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size):
    chunks = []
    bytes_received = 0

    while bytes_received < size:
        chunk = socket.recv(size - bytes_received)
        if not chunk:
            break

        chunks.append(chunk)
        bytes_received += len(chunk)
    return b''.join(chunks)[:bytes_received]



def send_all(socket: socket.socket, bytes):
    bytes_to_send = len(bytes)
    bytes_sent = 0
    while bytes_sent < bytes_to_send:
        n = socket.send(bytes[bytes_sent:])
        if n == -1:
            raise Exception("Socket error during write operation")
        bytes_sent += n
    return bytes_sent
